### PR TITLE
fix: alpha calculation when mean length exceeds target

### PR DESCRIPTION
**Pull request by dongcool**

### CODE REVIEW

File: contracts/phoenix-bonds/src/legacy.rs [hunk 1]
```diff
@@ -29,6 +31,74 @@ pub struct ContractV1_0_0 {
...
+        }
+    }
+}
+
+// ------ v1.1.0 ------
+// - Replace last_updated_at with exceeds_target_at in AccrualParameterV1_0_0
+
+#[near_bindgen]
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct ContractV1_1_0 {
```
Comment (dongcool) [inline contracts/phoenix-bonds/src/legacy.rs:81]: TODO migration function

File: contracts/phoenix-bonds/src/accrual.rs [hunk 2]
```diff
@@ -108,9 +108,16 @@ impl AccrualParameter {
         self.mean_length.insert(amount, ts); // this could only make mean length shorter
         let new_mean_length = self.mean_length.mean(ts);
 
-        if old_mean_length >= self.target_mean_length && new_mean_length < self.target_mean_length {
+        // if mean length drops below target, reset exceeds_target_at
+        if old_mean_length > self.target_mean_length && new_mean_length < self.target_mean_length {
```
Comment (linguists) [inline contracts/phoenix-bonds/src/accrual.rs:112]:
```suggestion
        if old_mean_length > self.target_mean_length && new_mean_length <= self.target_mean_length {
```

File: contracts/phoenix-bonds/src/accrual.rs [hunk 3]
```diff
@@ -121,12 +128,19 @@ impl AccrualParameter {
...
-                && new_mean_length >= self.target_mean_length)
-        {
+        // if mean length drops below target, reset exceeds_target_at
+        if old_mean_length > self.target_mean_length && new_mean_length < self.target_mean_length {
             self.alpha = alpha_before_removal;
-            self.last_updated_at = ts;
+            self.exceeds_target_at = 0;
+        } else if self.exceeds_target_at == 0 && new_mean_length > self.target_mean_length {
+            // if this action makes the mean length above target, then exceeds_target_at should be now.
+            if old_mean_length < self.target_mean_length {
```
Comment (linguists) [inline contracts/phoenix-bonds/src/accrual.rs:137]:
```suggestion
            if old_mean_length <= self.target_mean_length {
```

File: contracts/phoenix-bonds/src/accrual.rs [hunk 4]
```diff
@@ -121,12 +128,19 @@ impl AccrualParameter {
         self.mean_length.remove(amount, length, ts);
         let new_mean_length = self.mean_length.mean(ts);
 
-        if (old_mean_length >= self.target_mean_length && new_mean_length < self.target_mean_length)
-            || (old_mean_length < self.target_mean_length
-                && new_mean_length >= self.target_mean_length)
-        {
+        // if mean length drops below target, reset exceeds_target_at
+        if old_mean_length > self.target_mean_length && new_mean_length < self.target_mean_length {
```
Comment (linguists) [inline contracts/phoenix-bonds/src/accrual.rs:132]:
```suggestion
        if old_mean_length > self.target_mean_length && new_mean_length <= self.target_mean_length {
```

File: contracts/phoenix-bonds/src/accrual.rs (hunk 2 above)
Comment (linguists) [reply contracts/phoenix-bonds/src/accrual.rs:112]: When `new_mean_length <= self.target_mean_length` (including the case when the two are equal) , the alpha decreasing should be considered stopped I think.

**Review by linguists — CHANGES_REQUESTED** [review]

File: contracts/phoenix-bonds/src/legacy.rs [hunk 5]
```diff
@@ -1,14 +1,84 @@
...
+
+impl From<AccrualParameterV1_0_0> for AccrualParameter {
+    fn from(val: AccrualParameterV1_0_0) -> Self {
+        Self {
+            alpha: val.alpha,
+            min_alpha: val.min_alpha,
+            target_mean_length: val.target_mean_length,
+            adjust_interval: val.adjust_interval,
+            adjust_rate: val.adjust_rate,
+            exceeds_target_at: 0, // TODO
```
Comment (dongcool) [inline contracts/phoenix-bonds/src/legacy.rs:70]: if this is patched before the first time alpha should decrease then 0 should work

File: contracts/phoenix-bonds/src/legacy.rs [hunk 6]
```diff
@@ -1,14 +1,84 @@
...
+            mean_length: val.mean_length,
+        }
+    }
+}
+
+// ------ v1.1.0 ------
```
Comment (linguists) [inline contracts/phoenix-bonds/src/legacy.rs:76]: I suggest we use `v1.0.1` as a patch.

**Review by linguists — CHANGES_REQUESTED** [review]

File: contracts/phoenix-bonds/src/legacy.rs (hunk 5 above)
Comment (linguists) [reply contracts/phoenix-bonds/src/legacy.rs:70]: yes. this should work well. we can remove the `TODO` I think.

**Review by linguists — APPROVED** [review]

**Pull request merged**